### PR TITLE
[ORTTrainer] Fix to be compatible with onnxruntime 1.11.0

### PR DESCRIPTION
**Pull request by JingyaHuang**
# What does this PR do?
-  Fix `fix_atenops_to_gather` broken due to changed ops names.
-  Patch `Seq2SeqORTTrainer` to be compatible with `BART` model.
- Enable inference within PyTorch when ONNX Runtime doesn't work.

### CODE REVIEW

File: optimum/onnxruntime/trainer.py [hunk 1]
```diff
@@ -1194,7 +1229,7 @@ def compute_loss_ort(self, model, inputs, input_names, output_names, return_outp
 
         input_feed = dict(map(lambda input_name: (input_name, inputs[input_name].cpu().numpy()), input_names))
         outputs = self.infer_sess.run(output_names, input_feed)
-        outputs[0] = torch.Tensor(outputs[0])  # logits: `numpy.ndarray` -> tensor
+        outputs[0] = torch.Tensor(outputs[0])
```
Comment (echarlaix) [inline optimum/onnxruntime/trainer.py:1232]: Shouldn't you do this on the loss as [here](https://github.com/huggingface/optimum/blob/ort-train-11-compatibility/optimum/onnxruntime/trainer.py#L1242) you leave the possibility of outputs being a dictionnary

File: tests/onnxruntime/test_onnxruntime_train.py [hunk 2]
```diff
@@ -90,7 +86,8 @@ def compute_metrics(eval_pred):
                             return metric.compute(predictions=predictions, references=eval_pred.label_ids)
 
                         training_args = TrainingArguments(
-                            output_dir=tmp_dir,
+                            # output_dir=tmp_dir,
+                            output_dir="./results",
```
Comment (echarlaix) [inline tests/onnxruntime/test_onnxruntime_train.py:90]: Shouldn't it be `tmp_dir` ?

File: optimum/onnxruntime/trainer.py (hunk 1 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/trainer.py:1232]: Actually, the function `compute_loss_ort` will only be used for the ORT circumstance, otherwise `compute_loss` will be used. Thus the case of `outputs` being a dictionary does not really exist in that case. 
Indeed it is confusing because I didn't remove that cases of  `outputs` being a dictionary just 10 lines below...

File: tests/onnxruntime/test_onnxruntime_train.py (hunk 2 above)
Comment (JingyaHuang) [reply tests/onnxruntime/test_onnxruntime_train.py:90]: yep! Will correct that!

File: optimum/onnxruntime/trainer.py (hunk 1 above)
Comment (echarlaix) [reply optimum/onnxruntime/trainer.py:1232]: I think it would make sense to remove this case then, what do you think ?

Comment (JingyaHuang) [reply optimum/onnxruntime/trainer.py:1232]: Yep, done in the new commit!

File: optimum/onnxruntime/trainer.py [hunk 3]
```diff
@@ -657,16 +658,39 @@ def evaluate(
...
+                    prediction_loss_only=True if self.compute_metrics is None else None,
+                    ignore_keys=ignore_keys,
+                    metric_key_prefix=metric_key_prefix,
+                )
+            except Exception as error:
+                logger.warning(
+                    f"[ERROR!] Evaluation with ONNX Runtime is not available for {self.model.config.name_or_path} model. Remove `inference_with_ort` to evaluate within PyTorch."
+                )
+                logger.error(error)
+                raise
```
Comment (echarlaix) [inline optimum/onnxruntime/trainer.py:679]: Why not `raise error` ?

File: optimum/onnxruntime/trainer.py [hunk 4]
```diff
@@ -703,10 +731,34 @@ def predict(
...
+                    description="Prediction",
+                    ignore_keys=ignore_keys,
+                    metric_key_prefix=metric_key_prefix,
+                )
+            except Exception as error:
+                logger.warning(
+                    f"[ERROR!] Prediction with ONNX Runtime is not available with {self.model.config.name_or_path} model. Remove `inference_with_ort` to predict within PyTorch."
+                )
+                logger.error(error)
+                raise
```
Comment (echarlaix) [inline optimum/onnxruntime/trainer.py:749]: same comment

File: optimum/onnxruntime/trainer_seq2seq.py [hunk 5]
```diff
@@ -72,7 +74,17 @@ def evaluate(
         """
         self._max_length = max_length
         self._num_beams = num_beams
-        return super().evaluate(eval_dataset, ignore_keys=ignore_keys, metric_key_prefix=metric_key_prefix)
+        if self.args.predict_with_generate and inference_with_ort:
+            logger.error(
+                f"[ERROR!] Generate method is not available with prediction within ONNX Runtime. Remove `inference_with_ort` or `predict_with_generate`."
+            )
+            raise
```
Comment (echarlaix) [inline optimum/onnxruntime/trainer_seq2seq.py:81]:
```suggestion
            raise NotImplementedError("Generate method is not available with prediction within ONNX Runtime. Remove `inference_with_ort` or `predict_with_generate`.")
```

File: optimum/onnxruntime/trainer_seq2seq.py [hunk 6]
```diff
@@ -113,7 +126,17 @@ def predict(
         """
         self._max_length = max_length
         self._num_beams = num_beams
-        return super().predict(test_dataset, ignore_keys=ignore_keys, metric_key_prefix=metric_key_prefix)
+        if self.args.predict_with_generate and inference_with_ort:
+            logger.error(
+                f"[ERROR!] Generate method is not available with prediction within ONNX Runtime. Remove `inference_with_ort` or `predict_with_generate`."
+            )
+            raise
```
Comment (echarlaix) [inline optimum/onnxruntime/trainer_seq2seq.py:133]:
```suggestion
            raise NotImplementedError("Generate method is not available with prediction within ONNX Runtime. Remove `inference_with_ort` or `predict_with_generate`.")
```

File: optimum/onnxruntime/trainer.py (hunk 3 above)
Comment (JingyaHuang) [reply optimum/onnxruntime/trainer.py:679]: Actually here I will be more likely to put then in a reverse order, like:
 ```
logger.error(error)
logger.warning(
                    f"[ERROR!] Evaluation with ONNX Runtime is not available for {self.model.config.name_or_path} model. Remove `inference_with_ort` to evaluate within PyTorch."
                )
raise            
```
 
I want our warning to be in the last line so that it could be more clear for user to know what to do next. The original error could not normally be helpful for them.

File: optimum/onnxruntime/trainer.py [hunk 7]
```diff
@@ -1194,7 +1226,7 @@ def compute_loss_ort(self, model, inputs, input_names, output_names, return_outp
 
         input_feed = dict(map(lambda input_name: (input_name, inputs[input_name].cpu().numpy()), input_names))
         outputs = self.infer_sess.run(output_names, input_feed)
-        outputs[0] = torch.Tensor(outputs[0])  # logits: `numpy.ndarray` -> tensor
+        outputs[0] = torch.Tensor(outputs[0])
```
Comment (echarlaix) [inline optimum/onnxruntime/trainer.py:1229]: Should be removed.

File: optimum/onnxruntime/trainer.py [hunk 8]
```diff
@@ -1204,7 +1236,8 @@ def compute_loss_ort(self, model, inputs, input_names, output_names, return_outp
             loss = self.label_smoother(outputs, labels)
         else:
             # We don't use .loss here since the model may return tuples instead of ModelOutput.
-            loss = outputs["loss"] if isinstance(outputs, dict) else outputs[0]
+            # loss = outputs["loss"] if isinstance(outputs, dict) else outputs[0]
+            loss = outputs[0]
```
Comment (echarlaix) [inline optimum/onnxruntime/trainer.py:1240]: Should be removed as `outputs[0]` is not the loss.

**Review by echarlaix — APPROVED** [review]
